nomad/client_agent_endpoint.go

```diff
@@ -0,0 +1,237 @@
+package nomad
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+	"time"
+
+	log "github.com/hashicorp/go-hclog"
+	sframer "github.com/hashicorp/nomad/client/lib/streamframer"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/command/agent/monitor"
+	"github.com/hashicorp/nomad/helper"
+	"github.com/hashicorp/nomad/nomad/structs"
+
+	"github.com/ugorji/go/codec"
+)
+
+type Agent struct {
+	srv *Server
+}
+
+func (m *Agent) register() {
+	m.srv.streamingRpcs.Register("Agent.Monitor", m.monitor)
+}
+
+func (m *Agent) monitor(conn io.ReadWriteCloser) {
+	defer conn.Close()
+
+	// Decode args
+	var args cstructs.MonitorRequest
+	decoder := codec.NewDecoder(conn, structs.MsgpackHandle)
+	encoder := codec.NewEncoder(conn, structs.MsgpackHandle)
+
+	if err := decoder.Decode(&args); err != nil {
+		handleStreamResultError(err, helper.Int64ToPtr(500), encoder)
+		return
+	}
+
+	// Check node read permissions
+	if aclObj, err := m.srv.ResolveToken(args.AuthToken); err != nil {
+		handleStreamResultError(err, nil, encoder)
+		return
+	} else if aclObj != nil && !aclObj.AllowAgentRead() {
+		handleStreamResultError(structs.ErrPermissionDenied, helper.Int64ToPtr(403), encoder)
+		return
+	}
+
+	logLevel := log.LevelFromString(args.LogLevel)
+	if args.LogLevel == "" {
+		logLevel = log.LevelFromString("INFO")
+	}
+
+	if logLevel == log.NoLevel {
+		handleStreamResultError(errors.New("Unknown log level"), helper.Int64ToPtr(400), encoder)
+		return
+	}
+
+	// Targeting a client so forward the request
+	if args.NodeID != "" {
+		nodeID := args.NodeID
```

Comment:
maybe drop this block in a `func (m *Agent) forwardMonitor(...)` block for readability